libs/ssh/server.go

```diff
@@ -0,0 +1,322 @@
+	}
+}
+
+func (handler *sshHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	id := r.URL.Query().Get("id")
+	if id == "" {
+		http.Error(w, "Missing 'id' query parameter", http.StatusBadRequest)
+		return
+	}
+
+	if conn, exists := handler.Connections.Get(id); exists && conn != nil {
+		cmdio.LogString(ctx, fmt.Sprintf("Client with id %s is already connected", id))
+		err := conn.AcceptHandover(r.Context(), w, r)
+		if err != nil {
+			cmdio.LogError(ctx, fmt.Errorf("failed to accept handover: %v", err))
+			http.Error(w, "Handover failed", http.StatusInternalServerError)
+		} else {
+			cmdio.LogString(ctx, "Handover accepted for client with id "+id)
+		}
+		return
+	}
+
+	proxy := newProxyConnection(nil)
+	err := proxy.Accept(w, r)
+	if err != nil {
+		cmdio.LogError(ctx, fmt.Errorf("failed to upgrade to websockets: %v", err))
+		return
+	}
+	if !handler.Connections.TryAdd(id, proxy) {
+		err = proxy.Close()
+		if err != nil {
+			cmdio.LogError(ctx, fmt.Errorf("failed to close websocket: %v", err))
+		}
+		return
+	}
+	defer func() {
+		cmdio.LogString(ctx, "Closing WebSocket connection")
+		handler.Connections.Remove(id)
+		err := proxy.Close()
+		if err != nil {
+			cmdio.LogError(ctx, fmt.Errorf("failed to close websocket: %v", err))
+		}
+	}()
+
+	ctx, cancel := context.WithCancel(r.Context())
+	defer cancel()
+
+	g, gCtx := errgroup.WithContext(ctx)
+
+	cmdio.LogString(ctx, "New WebSocket connection, starting sshd process")
+
+	sshdCmd := createSSHDProcess(gCtx, handler.SSHDConfigPath)
+
+	sshdCmd.Stderr = os.Stderr
+
+	sshdStdin, err := sshdCmd.StdinPipe()
+	if err != nil {
+		cmdio.LogError(ctx, fmt.Errorf("failed to get stdin pipe: %v", err))
+		return
+	}
+	defer sshdStdin.Close()
+
+	sshdStdout, err := sshdCmd.StdoutPipe()
+	if err != nil {
+		cmdio.LogError(ctx, fmt.Errorf("failed to get stdout pipe: %v", err))
+		return
+	}
+	defer sshdStdout.Close()
+
+	g.Go(func() error {
+		return sshdCmd.Run()
+	})
+
+	g.Go(func() error {
+		return proxy.Start(gCtx, sshdStdout, sshdStdin)
+	})
+
+	if err := g.Wait(); err != nil {
+		cmdio.LogError(ctx, fmt.Errorf("SSHHandler error: %v", err))
+	}
+}
+
+func handleTimeout(ctx context.Context, timedOutChannel chan bool, shutdownDelay time.Duration) {
+	<-timedOutChannel
+	cmdio.LogString(ctx, fmt.Sprintf("No SSH clients for %v, shutting down...", shutdownDelay))
+	os.Exit(0)
+}
+
+func findAvailablePort(startPort, maxAttempts int) (int, error) {
+	for i := range maxAttempts {
+		port := startPort + i
+		addr := fmt.Sprintf(":%d", port)
+
+		listener, err := net.Listen("tcp", addr)
+		if err == nil {
+			listener.Close()
+			return port, nil
+		}
+	}
+	return 0, fmt.Errorf("no available port found after %d attempts starting from port %d", maxAttempts, startPort)
+}
+
+func saveWorkspaceMetadata(ctx context.Context, client *databricks.WorkspaceClient, version, clusterID string, port int) error {
+	metadataBytes, err := json.Marshal(PortMetadata{Port: port})
```

Comment:
The diagram in the first PR includes writing the username to this file. Is that out of date?